src/main/java/org/quiltmc/enigma_plugin/proposal/MappingMergePackageProposer.java

```diff
@@ -0,0 +1,397 @@
+import org.quiltmc.enigma.api.analysis.index.mapping.MappingsIndex;
+import org.quiltmc.enigma.api.analysis.index.mapping.PackageIndex;
+import org.quiltmc.enigma.api.translation.mapping.EntryMapping;
+import org.quiltmc.enigma.api.translation.mapping.EntryRemapper;
+import org.quiltmc.enigma.api.translation.mapping.tree.EntryTree;
+import org.quiltmc.enigma.api.translation.representation.entry.ClassEntry;
+import org.quiltmc.enigma.api.translation.representation.entry.Entry;
+import org.tinylog.Logger;
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * Proposes the packages from the mappings stored in {@link MappingMergeNameProposer} onto all top-level classes.
+ * These package names can be changed via overrides, which are a simple set of package names keyed by the versions from the mappings in {@link MappingMergeNameProposer}.
+ * This proposer must override the user-inputted mappings in order to work, and thus will generate mappings that take priority over manually input ones.
+ *
+ * <p>
+ *     This works via stripping out the class name and replacing the package dynamically after each rename.
+ *     For example, if a class is renamed {@code package/Class} and the name for the class in the mappings to merge is {@code net/minecraft/MojangClass},
+ *     the manually input class name will changed according to the mappings, resulting in {@code net/minecraft/Class}.
+ * </p>
+ * <p>
+ *     The override format is JSON, with a structure mimicking a package tree and entirely composed of one object type:
+ *     <pre><code>
+ *         [
+ *             {
+ *                 "obf": "com",
+ *                 "deobf": "",
+ *                 "children": [
+ *                     {
+ *                         "obf": "mojang",
+ *                         "deobf": "quiltmc",
+ *                         "children": []
+ *                     }
+ *                 ]
+ *             },
+ *             {
+ *                 "obf": "net",
+ *                 "deobf": "",
+ *                 "children": []
+ *             }
+ *         ]
+ *     </code></pre>
+ *
+ *     The elements of each package object are as follows:
+ *     <ul>
+ *         <li>
+ *             {@code obf}: the unqualified name of the package in the provided mappings
+ *         </li>
+ *         <li>
+ *             {@code deobf}: the unqualified, overridden name of the package
+ *         </li>
+ *         <li>
+ *             {@code children}: a list of child packages of this package.
+ *             The root of each override file is identical to one of these objects.
```

Comment:
this is a bit confusing imo, I'd say that root of the file is an array of top-level package objects before showing the example file